src/VisualStudio/Core/Def/Implementation/DesignerAttribute/DesignerAttributeIncrementalAnalyzer.cs

```diff
@@ -213,11 +213,23 @@ private async Task RegisterDesignerAttributeAsync(Document document, string desi
             var updateService = await GetUpdateServiceIfCpsProjectAsync(document.Project, cancellationToken).ConfigureAwait(false);
             if (updateService != null)
             {
-                // we track work by async token but doesn't explicitly wait for it. and update service is free-thread service,
-                // no need to switch to UI thread to use it
                 var asyncToken = _listener.BeginAsyncOperation("RegisterDesignerAttribute");
-                _ = updateService.SetProjectItemDesignerTypeAsync(document.FilePath, designerAttributeArgument)
-                                 .ReportNonFatalErrorAsync().CompletesAsyncOperation(asyncToken);
+                try
+                {
+                    // we track work by async token but doesn't explicitly wait for it. and update service is free-thread service,
+                    // no need to switch to UI thread to use it
+                    _ = updateService.SetProjectItemDesignerTypeAsync(document.FilePath, designerAttributeArgument)
+                                     .ReportNonFatalErrorAsync().CompletesAsyncOperation(asyncToken);
+                }
+                catch (ObjectDisposedException)
+                {
+                    // we might call update service after project is already removed and get object disposed exception.
+                    // we will catch the exception and ignore. this is wierd setup since we are catching exception from
+                    // async call, but until we have better way to handle this, this is what we have.
+                    // see this PR for more detail - https://github.com/dotnet/roslyn/pull/35383
+                    asyncToken.Dispose();
```

Comment:
since we're using try/catch, why not just have this be in `finally`?    with an `await` of `SetProjectItemDesignerTypeAsync` this would then be safe against all possible impls and would have the right catch and disposal semantics, right?

Reply:
if we use await, we don't need async token. @lifengl told me not to await 

at this point, I would like to see actual code.

Reply:
You can await it, as long as it doesn't affect you to process other files, and set sub type for them.  If you call multiple times quickly, we will batch them into one operation.  Otherwise, the project will be reevaluated after each change.